mne_connectivity/base.py

```diff
@@ -0,0 +1,172 @@
+                                   f'The current shape is {data.shape}.')
+
+        if not isinstance(data, np.ndarray):
+            raise TypeError('Connectivity data must be stored as a '
+                            'numpy array.')
+
+        if np.ndim(data) < 2 or np.ndim(data) > 4:
+            raise ValueError(f'Connectivity data that is passed '
+                             f'must be either 2D, 3D, or 4D, where the '
+                             f'last axis is time if 3D or 4D and the '
+                             f'3rd axis is frequency if 4D. The shape '
+                             f'passed was {data.shape}.')
+        return data
+
+    @property
+    def shape(self):
+        return self.data.shape
+
+    @property
+    def epoch_axis(self):
+        return None
+
+    @property
+    def frequency_axis(self):
+        raise NotImplementedError(
+            'The frequency axis is not defined here. Please '
+            'double check that you are using the right type of '
+            'connectivity container.')
+
+    @property
+    def time_axis(self):
+        raise NotImplementedError(
+            'The time axis is not defined here. Please '
+            'double check that you are using the right type of '
+            'connectivity container.')
+
+    def get_data(self, return_nans=False):
+        if return_nans:
+            return self.data
+
+        if self.indices is not None:
+            row_idx, col_idx = self.indices
+            return self.data[row_idx, col_idx, ...]
+        else:
+            return self.data
+
+    def plot_circle(self):
+        pass
+
+    def plot_matrix(self):
+        pass
+
+    def plot_3d(self):
+        pass
+
+    def save(self, fname):
+        pass
+
+
+class SpectralConnectivity(_Connectivity):
+    def __init__(self, data, names, freqs, indices=None, method=None,
+                 spec_method=None, n_epochs=None):
+        super().__init__(data, names, method, indices)
+
+        self.freqs = freqs
+        self.spec_method = spec_method
+        self.n_epochs = n_epochs
+
+    @property
+    def frequency_axis(self):
+        return 2
+
+
+class TemporalConnectivity(_Connectivity):
+    def __init__(self, data, names, times, indices=None,
+                 method=None, n_epochs=None):
+        super().__init__(data, names, method, indices)
+
+        self.times = times
+        self.n_epochs = n_epochs
+
+    @property
+    def time_axis(self):
+        return 2
+
+
+class SpectroTemporalConnectivity(_Connectivity):
+    def __init__(self, data, names, freqs, times, indices=None,
+                 method=None, spec_method=None, n_epochs=None):
+        super().__init__(data, names, method, indices)
+
+        self.freqs = freqs
+        self.spec_method = spec_method
+        self.times = times
+        self.n_epochs = n_epochs
+
+    @property
+    def frequency_axis(self):
+        return 2
+
+    @property
+    def time_axis(self):
+        return 3
```

Comment:
If we can standardize on time axis being `-1` and freq axis being `-2` then we don't need these.

Also if we really want axis labels then we could just have `def axis_labels(self): return ('epochs', 'frequencies', 'time')` or so. This is starting to feel a little bit like XArray territory...

@agramfort you've opposed using XArray in the past, WDYT about trying it out in mne-connectivity? We could have `object.data` return an `ndarray` cast of the internally stored xarray, and have `object.xarray` for example give the root xarray object.

Reply:
However, sometimes the frequency axis is not there, and sometimes the time-axis is not there?

Reply:
Having vs not having properties is a bit ugly because it leads to code like `isinstance(...)` or `hasattr(...)`, might be nicer just to have a `.dimensions` property that is a tuple of str. Then if you want the axis number you can always do `obj.dimensions.index('time')` for example. But I would look at the xarray API for inspiration/naming here

Reply:
Makes sense! For my reference: http://xarray.pydata.org/en/stable/user-guide/terminology.html

looks like: `conn.dims` will show `('node', 'node', 'time', ...)`

Reply:
Sounds good, though we should probably call it `node_from` and `node_to` to indicate potential directionality

Reply:
Ooo yeah that sounds way better!

Reply:
no objection for xarray here. I think it's a good fit for the usecase here.

Reply:
Xarray has officially been integrated... And it's awesome! 

`dims`, `attrs`, `coords` are just nice semantic features of each data axis.